### PR TITLE
[MM-43850] Separate leave_team events for associated user and team

### PR DESCRIPTION
**Pull request by plant99**
#### Summary

`leave_team` event fails to reach user in some cases. This PR configures separates the events going to the associated user and rest of the team, and uses `SendType = model.ClusterSendReliable` for the user event.

Note: I was unable to reproduce this, but this is a general improvement nonetheless. 

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-43850
#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Comment by agnivade** [comment]
I don't think this is the right fix. As Daniel pointed out, this seems to be a race condition when the websocket event gets processed slightly delayed, and by that time the user session is updated to reflect that the user is no longer a team member. In that case, the event won't be sent.

The websocket events are processed in a buffered channel. So publishing an event does not mean it gets sent immediately. And if an event is team scoped, for every event, it will check if the user is a member of that team or not: https://github.com/mattermost/mattermost-server/blob/5a4cac43a8e2b732dc5985706daf66a16e59a2f0/app/platform/web_conn.go#L796-L798.

So this means, if there were lots of events in the buffered channel, and by the time, the removeMember event is processed, the user is actually deleted from the DB, this method will return false and no event will be sent out.

What you can do is try to reorder the code to have the team member deleted in the DB first, and the session invalidated, and then send the ws event. For ex: https://github.com/mattermost/mattermost-server/blob/5a4cac43a8e2b732dc5985706daf66a16e59a2f0/app/team.go#L1323 bring this line above `RemoveTeamMember` and see what happens. If you don't see the event, then that's what's happening.

**Comment by plant99** [comment]
> I don't think this is the right fix. As Daniel pointed out, this seems to be a race condition when the websocket event gets processed slightly delayed, and by that time the user session is updated to reflect that the user is no longer a team member. In that case, the event won't be sent.
> 
> The websocket events are processed in a buffered channel. So publishing an event does not mean it gets sent immediately. And if an event is team scoped, for every event, it will check if the user is a member of that team or not:
> 
> https://github.com/mattermost/mattermost-server/blob/5a4cac43a8e2b732dc5985706daf66a16e59a2f0/app/platform/web_conn.go#L796-L798
> 
> .
> So this means, if there were lots of events in the buffered channel, and by the time, the removeMember event is processed, the user is actually deleted from the DB, this method will return false and no event will be sent out.
> 


Thanks @agnivade, for the detailed explanation, really helped clarify things more.
> What you can do is try to reorder the code to have the team member deleted in the DB first, and the session invalidated, and then send the ws event. For ex:
> 
> https://github.com/mattermost/mattermost-server/blob/5a4cac43a8e2b732dc5985706daf66a16e59a2f0/app/team.go#L1323
> 
> bring this line above `RemoveTeamMember` and see what happens. If you don't see the event, then that's what's happening.

It seems like we've 'accidentally' created the right solution. I did this exercise, and since this patch generates a [new event only for user](https://github.com/mattermost/mattermost-server/pull/21231/files#diff-a1dfe5f32427ac215ba8a21b24fd69360f2c9c7f1992928f5f47a85c2b68ea7bR207) and has `msg.GetBroadcast().TeamId = ""`, it goes through.

I've removed the unnecessary `ReliableClusterSend` modification.

File: app/teams/teams.go [hunk 1]
```diff
@@ -192,11 +192,28 @@ func (ts *TeamService) JoinUserToTeam(team *model.Team, user *model.User) (*mode
...
+	messageTeam.Add("team_id", teamMember.TeamId)
+	ts.wh.Publish(messageTeam)
+
+	// message for teamMember.UserId
+	messageUser := model.NewWebSocketEvent(model.WebsocketEventLeaveTeam, "", "", teamMember.UserId, nil, "")
+	messageUser.Add("user_id", teamMember.UserId)
+	messageUser.Add("team_id", teamMember.TeamId)
+
+	broadcast := messageUser.GetBroadcast()
+	messageUser.SetBroadcast(broadcast)
```
Comment (agnivade) [inline app/teams/teams.go:212]: Need to remove these 2 lines as well

Comment (plant99) [reply app/teams/teams.go]: oops! thanks for pointing it out

**Review by agnivade — APPROVED** [review]
Nice work. I am also wondering if this applies to channel_leave event as well. Could you give that a look when you get a chance?

**Comment by plant99** [comment]
> Nice work. I am also wondering if this applies to channel_leave event as well. Could you give that a look when you get a chance?

I'll create a ticket for that as well.

**Review by mkraft — APPROVED** [review]

**Comment by plant99** [comment]
> > Nice work. I am also wondering if this applies to channel_leave event as well. Could you give that a look when you get a chance?
> 
> I'll create a ticket for that as well.

@agnivade https://github.com/mattermost/mattermost-server/blob/master/app/channel.go#L2505-L2514 not creating a ticket for this fyi, since it's already implemented. Thanks for raising this, at least we could double check!

**Review by furqanmlk — APPROVED** [review]
Working fine

**Pull request merged**